src/update_notifier.rs

```diff
@@ -3,117 +3,231 @@
-            if now - last_checked_time >= time_to_check {
-                config::lock_background_process()?;
-                // lock and check for lock
-                std::process::Command::new("wapm")
-                    .arg("run-background-update-check")
-                    .stdout(std::process::Stdio::null())
-                    .stderr(std::process::Stdio::null())
-                    .spawn()
-                    .ok()?;
+#[derive(Deserialize, Serialize, Debug, PartialEq)]
+pub struct WapmUpdate {
+    /// The data related to the last check on the Github Registry
+    pub last_check: Option<WapmLastCheck>,
+    /// The time when wapm last trigger the notification
+    pub last_notified: Option<DateTime<Utc>>,
+}
+
+impl WapmUpdate {
+    fn default() -> Self {
+        Self {
+            last_check: None,
+            last_notified: None,
+        }
+    }
+    fn load() -> Result<Self, String> {
+        let path = get_wapm_update_file_path();
+        let json_file = File::open(path).map_err(|err| err.to_string())?;
+        let wasm_update: WapmUpdate =
+            serde_json::from_reader(json_file).map_err(|err| err.to_string())?;
+        Ok(wasm_update)
+    }
+    fn load_or_default() -> Self {
+        Self::load().unwrap_or_else(|_e| Self::default())
+    }
+    fn save(&self) -> Result<(), String> {
+        let path = get_wapm_update_file_path();
+        let json_file = File::create(path).map_err(|err| err.to_string())?;
+        serde_json::to_writer(json_file, &self).map_err(|err| err.to_string())?;
+        Ok(())
+    }
+    fn set_last_check(&mut self, version: String) {
+        let now = Utc::now();
+        self.last_check = Some(WapmLastCheck {
+            timestamp: now,
+            version: version,
+        });
+    }
+    fn should_trigger_check(&self) -> bool {
+        match self.last_check {
+            Some(ref last_check) => {
+                let now = Utc::now();
+                let time_to_check: time::Duration = time::Duration::from_std(
+                    std::time::Duration::from_secs(CHECK_DURATION_IN_SECONDS),
+                )
+                .unwrap();
+                now - last_check.timestamp >= time_to_check
             }
+            None => true,
         }
     }
+    fn maybe_print_notification(&mut self) -> Result<(), String> {
+        let last_check = self.last_check.as_ref();
+        match last_check {
+            None => Ok(()),
+            Some(last_check) => {
+                let now = Utc::now();
+                if let Some(last_notified) = self.last_notified {
+                    let time_to_check: time::Duration = time::Duration::from_std(
+                        std::time::Duration::from_secs(WAPM_NOTIFICATION_WINDOW),
+                    )
+                    .unwrap();
+                    if now - last_notified < time_to_check {
+                        return Ok(());
```

Comment:
this is duplicating logic from `should_trigger_check`, we can just call it here

Reply:
I think it was good to have the two concepts separate, as the windows for each might be different:
* When do we schedule a new check (each 24 hours)
* When we show a notification (each 2 hours)